toolchain/lower/file_context.h

```diff
@@ -219,6 +294,34 @@ class FileContext {
 
   // Global format string for `printf.int.format` used by the PrintInt builtin.
   llvm::Value* printf_int_format_string_ = nullptr;
+
+  // For a generic function, keep track of the specifics for which LLVM
+  // function declarations were created. Those can be retrieved then via
+  // from specific_functions_, via specific_functions_[specific_id.index].
+  // We resize this to the correct size. Indexed by generic_id.index.
+  llvm::SmallVector<llvm::SmallVector<SemIR::SpecificId>, 0> lowered_specifics_;
+
+  // For specifics that exist in lowered_specifics, a hash of their function
+  // type information: return and parameter types. We resize this to the
+  // correct size. Indexed by specific_id.index.
+  // TODO: Hashing all members of FunctionTypeInfo may not be necessary.
+  llvm::SmallVector<llvm::BLAKE3Result<32UL>, 0>
+      lowered_specifics_type_fingerprint_;
+
+  // This is initialized and populated while lowering a specific.
+  // We resize this to the correct size. Indexed by specific_id.index.
+  llvm::SmallVector<SpecificFunctionFingerprint, 0>
+      lowered_specific_fingerprint_;
+
+  // Equivalent specifics found: for each specific point to the
+  // canonical equivalent specific
+  llvm::SmallVector<SemIR::SpecificId> equivalent_specifics_;
+  // Non-equivalent specifics found.
+  Set<std::pair<SemIR::SpecificId, SemIR::SpecificId>>
+      non_equivalent_specifics_;
+  // Track whether a specific was replaced by another. We resize it to the
+  // correct size and initialize with all false. Indexed by specific_id.index.
+  llvm::SmallVector<bool, 0> is_replaced_specific_;
```

Comment:
Had you considered extracting the coalescing logic and data out to its own type, rather than directly on `FileContext`? i.e., similar to how we have `DeclNameStack` in `check`, just moving this to an object and having `FileContext` hold an instance of that object. There's a fair amount of it now, and I'm wondering if it'd make it easier (in part, I'm just expecting the logic to grow as more edge-cases are handled).

Reply:
Note similarly for function_context.h, if you moved this stuff out, you could also have something like a `FunctionFingerprinter` with the various `Add` methods on it in the same file. i.e., to have the specific fingerprinting logic collected together.

Reply:
Leaving this refactoring as a separate part. Yes, it should be extracted from FileContext.